Prove regression from #1893

dhensby opened the pull request:
see #1898 

The test that was merged with #1893 didn't actually test that lines after a user statement weren't blank as PHPCS would determine this to be the end of the file due to lack of class declaration or other valid code lines.

dhensby commented:
Right, I've pushed a fix that allows a trailing comment on the same line as a `use` statement.

I'd have thought this should be a config option, but as #1893 was merged without a new config I'd assume it's intended that trailing comments are allowed.

iammattcoleman commented on src/Standards/PSR2/Sniffs/Namespaces/UseDeclarationSniff.php:
```diff
@@ -137,12 +137,19 @@ public function process(File $phpcsFile, $stackPtr)
 
-        $end = $phpcsFile->findNext(T_SEMICOLON, ($stackPtr + 1));
+        $end         = $phpcsFile->findNext(T_SEMICOLON, ($stackPtr + 1));
+        $nextComment = $phpcsFile->findNext(T_COMMENT, ($end + 1));
```
Should this be `Tokens::$emptyTokens` to allow for `phpcs:ignore`, etc.?

dhensby replied:
Side note: What we are checking for here is the next inline comment, if that's on the same line as the `use` declaration then that can be classed as the "end" of the `use` declaration instead of the semicolon.

As I see it: Allowing for any "empty token" would mean a `/*` could be allowed (which it shouldn't be IMO) and other such tokens that shouldn't or couldn't be there.

I'm happy to be told that it **should** use `Tokens::$emptyTokens` I'm just of the opinion that the only thing that can come next should be an inline comment.

iammattcoleman replied:
`3.2.0` changed how overrides are annotated, which added several tokens for comments containing `phpcs:` overrides: https://github.com/squizlabs/PHP_CodeSniffer/releases/tag/3.2.0

My PR changed `T_COMMENT` to `Tokens::$emptyTokens` in order to allow the `T_PHPCS_*` tokens. That also allows `T_DOC_COMMENT_*`, which seemed safe but isn't 100% correct in hindsight. I think this should be an array with `T_COMMENT` and `T_PHPCS_*`.

@jrfnl Does that sound right to you?

jrfnl replied:
I'll have a look at this PR later today.

dhensby commented:
@iammattcoleman i've quickly tried to add `Tokens::$emptyTokens` and a while loop to get this working but I'm now getting other failures:

```
PHPUnit 5.7.27 by Sebastian Bergmann and contributors.

...............................................................  63 / 294 ( 21%)
............................................................... 126 / 294 ( 42%)
............................................................... 189 / 294 ( 64%)
.F.............................................SS.............. 252 / 294 ( 85%)
.......................................S..                      294 / 294 (100%)

Tests generated 618 unique error codes; 350 were fixable (56.63%)

Time: 27.52 seconds, Memory: 38.00MB

There was 1 failure:

1) PHP_CodeSniffer\Standards\PSR2\Tests\Namespaces\UseDeclarationUnitTest::testSniff
[LINE 12] Expected 0 error(s) in UseDeclarationUnitTest.1.inc but found 1 error(s). The error(s) found were:
 -> There must be one blank line after the last USE statement; 0 found; (PSR2.Namespaces.UseDeclaration.SpaceAfterLastUse)
[LINE 2] Expected 1 error(s) in UseDeclarationUnitTest.10.inc but found 0 error(s).
[LINE 10] Expected 2 error(s) in UseDeclarationUnitTest.2.inc but found 1 error(s). The error(s) found were:
 -> USE declarations must go after the first namespace declaration (PSR2.Namespaces.UseDeclaration.UseAfterNamespace)
Fixed version of UseDeclarationUnitTest.2.inc does not match expected version in UseDeclarationUnitTest.2.inc.fixed; the diff is
--- src/Standards/PSR2/Tests/Namespaces/UseDeclarationUnitTest.2.inc.fixed
+++ PHP_CodeSniffer
@@ -10,6 +10,7 @@

 use ArrayObject;

+
 $foo = 'bar';

 ?>
Fixed version of UseDeclarationUnitTest.3.inc does not match expected version in UseDeclarationUnitTest.3.inc.fixed; the diff is
--- src/Standards/PSR2/Tests/Namespaces/UseDeclarationUnitTest.3.inc.fixed
+++ PHP_CodeSniffer
@@ -3,8 +3,9 @@

 use someNS\A;
 use someNS\B;
+class

-class Bug
+Bug
 {
     public function __construct()
     {
[LINE 2] Expected 0 error(s) in UseDeclarationUnitTest.4.inc but found 1 error(s). The error(s) found were:
 -> There must be one blank line after the last USE statement; 0 found; (PSR2.Namespaces.UseDeclaration.SpaceAfterLastUse)
[LINE 19] Expected 1 error(s) in UseDeclarationUnitTest.5.inc but found 2 error(s). The error(s) found were:
 -> There must be one USE keyword per declaration (PSR2.Namespaces.UseDeclaration.MultipleDeclarations)
 -> There must be one blank line after the last USE statement; 0 found; (PSR2.Namespaces.UseDeclaration.SpaceAfterLastUse)
Fixed version of UseDeclarationUnitTest.5.inc does not match expected version in UseDeclarationUnitTest.5.inc.fixed; the diff is
--- src/Standards/PSR2/Tests/Namespaces/UseDeclarationUnitTest.5.inc.fixed
+++ PHP_CodeSniffer
@@ -27,6 +27,7 @@
 use function foo\math\cos;
 use function foo\math\cosh;

+
 class PHP7 {

 }
[LINE 2] Expected 0 error(s) in UseDeclarationUnitTest.8.inc but found 1 error(s). The error(s) found were:
 -> There must be one blank line after the last USE statement; 0 found; (PSR2.Namespaces.UseDeclaration.SpaceAfterLastUse)
[LINE 2] Expected 0 error(s) in UseDeclarationUnitTest.9.inc but found 1 error(s). The error(s) found were:
 -> There must be one blank line after the last USE statement; 0 found; (PSR2.Namespaces.UseDeclaration.SpaceAfterLastUse)

/vagrant/www/vendor/squizlabs/php_codesniffer/tests/Standards/AbstractSniffUnitTest.php:205
/vagrant/www/vendor/squizlabs/php_codesniffer/tests/TestSuite.php:28

FAILURES!
Tests: 294, Assertions: 302, Failures: 1, Skipped: 3.
```

dhensby commented:
Gah, I see my problem but don't have time to fix right now. We need the `$nextComment` from the last previous iteration

dhensby commented:
OK - I've pushed in another commit that uses `Tokens::$emptyTokens` and passes the tests. Maintainers have push rights to the branch so can reset that commit out if it's decided it's not needed.

iammattcoleman commented:
@gsherwood @jrfnl This should probably be included in 3.2.3.

dhensby commented:
> @gsherwood @jrfnl This should probably be included in 3.2.3.

Yep, unless you want 3.2.3 to be broken :)

iammattcoleman commented:
@gsherwood I'm currently helping with the update to PHP 7.2 for Ubuntu 18.04. Currently, PHP_CodeSniffer 3.2.2 is slated to be included. I've let them know that there are some important fixes coming soon, but I felt you should know that they're currently packaging 3.2.2 along with its known issues.

jrfnl commented:
As per @iammattcoleman @dhensby request, I've had a look at this PR. I believe the logic can be simplified slightly by just checking for either next statement or the start of the next line, whichever comes first and not focusing completely on "next comment".

I've just send in a PR to @dhensby's repo with an optional suggestion for amending this PR.
See https://github.com/dhensby/PHP_CodeSniffer/pull/1

jrfnl commented:
Since the original downstream PR to amend this PR, @iammattcoleman and me have identified some additional issues and the downstream PR has been updated to allow for problems in the fixer when encountering various combinations of trailing comments after `use` statements.
So my downstream PR is no longer _optional_.

dhensby commented:
I've merged the downstream PR - all looks good - thanks for the thorough investigation

jrfnl commented:
@dhensby Thanks for working with me to get this sorted.

dhensby commented:
@jrfnl no problem! Thanks for the help :)

gsherwood commented:
@dhensby Thanks for finding and fixing this before release. Very much appreciated. Thanks to those who reviewed as well.
